src/reanimated2/component/FlatList.tsx

```diff
@@ -60,48 +60,58 @@ interface AnimatedFlatListComplement<T> extends FlatList<T> {
   getNode(): FlatList<T>;
 }
 
-export const ReanimatedFlatList = forwardRef(
-  (
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    props: ReanimatedFlatListPropsWithLayout<any>,
-    ref: ForwardedRef<FlatList>
-  ) => {
-    const { itemLayoutAnimation, skipEnteringExitingAnimations, ...restProps } =
-      props;
+// We need explicit any here, because this is the exact same type that is used in React Native types
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const FlatListForwardRefRender = function <Item>(
+  props: ReanimatedFlatListPropsWithLayout<Item>,
+  ref: ForwardedRef<FlatList>
+) {
+  const { itemLayoutAnimation, skipEnteringExitingAnimations, ...restProps } =
+    props;
 
-    // Set default scrollEventThrottle, because user expects
-    // to have continuous scroll events and
-    // react-native defaults it to 50 for FlatLists.
-    // We set it to 1 so we have peace until
-    // there are 960 fps screens.
-    if (!('scrollEventThrottle' in restProps)) {
-      restProps.scrollEventThrottle = 1;
-    }
+  // Set default scrollEventThrottle, because user expects
+  // to have continuous scroll events and
+  // react-native defaults it to 50 for FlatLists.
+  // We set it to 1, so we have peace until
+  // there are 960 fps screens.
+  if (!('scrollEventThrottle' in restProps)) {
+    restProps.scrollEventThrottle = 1;
+  }
 
-    const CellRendererComponent = React.useMemo(
-      () => createCellRendererComponent(itemLayoutAnimation),
-      []
-    );
+  const CellRendererComponent = React.useMemo(
+    () => createCellRendererComponent(itemLayoutAnimation),
+    []
+  );
 
-    const animatedFlatList = (
-      <AnimatedFlatList
-        ref={ref}
-        {...restProps}
-        CellRendererComponent={CellRendererComponent}
-      />
-    );
+  const animatedFlatList = (
+    // @ts-expect-error In its current type state, createAnimatedComponent cannot create generic components.
+    <AnimatedFlatList
+      ref={ref}
+      {...restProps}
+      CellRendererComponent={CellRendererComponent}
+    />
+  );
+
+  if (skipEnteringExitingAnimations === undefined) {
+    return animatedFlatList;
+  }
 
-    if (skipEnteringExitingAnimations === undefined) {
-      return animatedFlatList;
-    }
+  return (
+    <LayoutAnimationConfig skipEntering skipExiting>
+      {animatedFlatList}
+    </LayoutAnimationConfig>
+  );
+};
 
-    return (
-      <LayoutAnimationConfig skipEntering skipExiting>
-        {animatedFlatList}
-      </LayoutAnimationConfig>
-    );
+export const ReanimatedFlatList = forwardRef(FlatListForwardRefRender) as <
+  // We need explicit any here, because this is the exact same type that is used in React Native types.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ItemT = any
+>(
+  props: ReanimatedFlatListPropsWithLayout<ItemT> & {
+    ref?: ForwardedRef<FlatList>;
   }
-);
+) => React.ReactElement;
```

Comment:
This makes me curious, are we certain that this type of general type here won't cause any issues?

Reply:
Hey! Let me look into that!

Reply:
@tjzel I wasn't able to find a more elegant solution. If you have some ideas, please share them with me.
To back using it like it is I can share the function call type declaration in the ExoticComponent interface. It also returns a `ReactElement`. But for sure there are parts of that missing.
<img width="561" alt="CleanShot 2023-12-28 at 19 06 51@2x" src="https://github.com/software-mansion/react-native-reanimated/assets/16062886/929b0194-69d4-484e-892d-3a82097a4994">

In my solution which I have in a patch already I was using below sources:
- https://fettblog.eu/typescript-react-generic-forward-refs/
- https://stackoverflow.com/questions/58469229/react-with-typescript-generics-while-using-react-forwardref

Reply:
Also https://github.com/Microsoft/TypeScript/issues/17574

Reply:
Hey, that's some amazing research there, very interesting reads! (it's tempting to request more changes just to read some more 😄)

Let's merge this bad boy